user-groups: Sort groups into editable and non-editable.

Kumar-Tarun opened the pull request:
This PR sorts user groups into 'Your groups' and 'Other groups' as -
![17](https://user-images.githubusercontent.com/23361315/37862504-d48621d8-2f73-11e8-95c2-9aebf566db57.PNG)

![19](https://user-images.githubusercontent.com/23361315/37862507-db411848-2f73-11e8-8c9a-3f6a9a226d6a.PNG)

timabbott commented on static/js/settings_user_groups.js:
```diff
@@ -29,7 +29,31 @@ exports.populate_user_groups = function () {
+        user_groups_section.append('<h3>Your groups</h3>');
+    } else if (user_groups_array.length > 0) {
+        user_groups_section.append('<h3>Your groups</h3><p>You are not a part of any groups.</p>');
+    }
```
We need these strings to be tagged for translation.

Kumar-Tarun commented:
@timabbott I've updated the PR. You may check!

timabbott commented on tools/lib/capitalization.py:
```diff
@@ -77,6 +77,7 @@
      '<a href="/integrations" target="_blank">Integrations</a> or '
      '<a href="/api" target="_blank">API</a> documentation?'),
     r'Most stream administration is done on the <a href="/#streams">Streams page</a>.',
+    r'<h3>Your groups</h3><p>You are not a part of any groups.</p',
```
This is wrong -- we should instead break that into 2 translatable strings, one for each block.

Kumar-Tarun replied:
Done!

Kumar-Tarun replied:
I am also thinking of refactoring this with work for issue #8802, it will make the whole thing clean!

timabbott commented:
@Kumar-Tarun you need to work on making your changes comprehensible.  E.g. why did you move `test_with_external_user` to a different section of the test file?  The tests pass without that change, and it makes the diff much harder to understand.

timabbott commented:
One thing I notice seems wrong at the end of this branch:
* An organization administrator sees groups they're not in under "Your groups" -- I think "Your groups" should only be groups you're actually in.

I think in this you'd benefit a lot from doing finer-grained changes that each do just 1 thing and not stacking refactoring on top of features (e.g. it'd be a lot clearer if you did a couple commits at the bottom of the branch that switched to the new non-editable pill template, and THEN worked on the sorting stuff.)

Kumar-Tarun commented:
@timabbott Thanks for this detailed review! Actually I got too eager to work on 'cleaning up' part, that I decided to work on that on top of the sorting commit, (without that being merged).
Regarding the move of that test function, you're right, while making changes I forgot to put it back in its original position. I am really sorry for all these mistakes. 
One thing I think here is since sorting and refactoring are related (and latter depends on former), can you merge the 1st commit of sorting part (I know its wrong regarding to that admin case, but I've rectified that in the clean-up part). Then we can have the clean up part following that.

Kumar-Tarun commented on static/js/settings_user_groups.js:
```diff
@@ -14,29 +14,67 @@ exports.reload = function () {
+            name: data.name,
+            id: data.id,
+            description: data.description,
+            is_non_member_first: editable_groups.indexOf(data) === editable_group_length,
```
I could have done this in JS as well without passing that to template, but it was making the tests quite complicated, this condition will hold in case of admins only, when there exists a group of which admin is not a member. Further, I removed 'user_group' context that was being sent as arguments to templates for two reasons - 
1) To avoid linting errors
2) It was not adding any special information.
Hoping that makes some sense. :)

Kumar-Tarun commented on static/templates/admin_user_group_list.handlebars:
```diff
@@ -1,4 +1,6 @@
-{{#with user_group}}
```
Again, this has been done to avoid linting errors.

Kumar-Tarun commented:
I've cleaned the PR. Some changes I want to make note of-
1. I squashed the sorting commit and clean up commit, since they were interrelated, and also sorting was the core logic behind the cleaning part, so I think it should be better to have that in one commit.
2. Now overall logic is pretty simple, we sort groups into editable and non-editable which has two sub-cases:
1. for normal users, editable part gets rendered through `admin_user_group_list`, and non-editable part gets rendered through `non_editable_user_group`.
2. for admins we have all groups in `editable` array (so gets rendered through `admin_user_group_list`), sorted by first editable and non-editable, and we just add one condition to have `non-editable` (meaning not member of) groups to be under `Other groups`.

So, the bottom-line is `editable` means which can be edited and `non-editable` means which cannot be edited. :)

timabbott commented:
Closing as we are replacing this with a new "user groups" panel similar to "stream settings".
